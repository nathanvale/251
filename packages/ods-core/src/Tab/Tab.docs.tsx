import React from "react";
import { ComponentDocs } from "@origin-digital/ods-types";
import { IconElectricity } from "@origin-digital/ods-icons";
import { Tab } from "./Tab";

export const docs: ComponentDocs = {
  category: "Content",
  componentName: "Tab",
  description: `An individual Tab, only useful as the child of the Tabs component.`,
  migrationGuide: false,
  propDescriptions: {
    value: "This tab is selected when this value matches the parent Tabs value",
    label: "Text label for the tab",
    icon: "Optional icon to be displayed to the left of the label",
    children:
      "The content to be shown when the tab is selected (handled by the Tabs component)",
    muiProps: "Allows access to the MUI component API under the hood",
  },
  examples: {
    default: {
      Code: () => <Tab data-id="tab1" label="Electricity" />,
    },
    additional: [
      {
        label: "Icon tab",
        description: "Tab with an icon on it",
        Code: () => (
          <Tab
            data-id="tab1"
            label="Electricity"
            icon={<IconElectricity tone="neutral" />}
          />
        ),
      },
    ],
  },
  snippets: [],
};
