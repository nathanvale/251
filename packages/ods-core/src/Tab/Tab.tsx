import React from "react";
import {
  Tab as MuiTab,
  TabProps as MuiTabProps,
  makeStyles,
  Theme,
  fade,
} from "@material-ui/core";
import { ComponentBaseProps } from "@origin-digital/ods-types";
import { useTracking } from "../TrackingProvider/useTracking";
import { Box } from "../Box";

export interface TabProps extends ComponentBaseProps {
  label: string | React.ReactNode;
  icon?: React.ReactNode;
  value?: string | number;
  muiProps?: MuiTabProps;
}

const useTabStyles = makeStyles(
  (theme: Theme) => ({
    root: {
      color: theme.palette.secondaryB.main,
      opacity: 1,
      textTransform: "none",
      padding: `${theme.spacing(3)}px ${theme.spacing(4)}px`,
      fontSize: `${theme.typography.body1.fontSize}px`,
      "&:hover": {
        color: theme.palette.grey[600],
        backgroundColor: theme.palette.grey[100],
      },
      "&$selected .MuiTab-wrapper": {
        fontWeight: theme.typography.fontWeightMedium,
      },
      "&$selected svg": {
        fill: theme.palette.secondary.main,
        color: theme.palette.secondary.main,
      },
      "&$selected:hover": {
        backgroundColor: fade(theme.palette.secondary.main, 0.08),
      },
      "&:focus": {
        color: theme.palette.grey[600],
        backgroundColor: fade(theme.palette.secondary.main, 0.16),
      },
      "&:active": {
        color: theme.palette.grey[600],
        backgroundColor: fade(theme.palette.secondary.main, 0.24),
      },
    },
    selected: {},
  }),
  { classNamePrefix: "tab" }
);

export const Tab = ({
  icon,
  label,
  "data-id": dataId = "tab",
  muiProps,
  value,
  ...others
}: TabProps) => {
  const classes = useTabStyles();
  const { onClickCapture, ref } = useTracking({
    children: label,
    "data-id": dataId,
    type: Tab.displayName,
    postClickState: "selected",
  });
  return (
    <MuiTab
      disableRipple
      data-id={dataId}
      classes={classes}
      onClickCapture={onClickCapture}
      ref={ref}
      label={
        <Box display="flex" flexDirection="row" justifyContent="center">
          {icon && <Box marginRight="xsmall">{icon}</Box>}
          {label}
        </Box>
      }
      value={value}
      id={`tab-${value}`}
      aria-controls={`tabpanel-${value}`}
      {...muiProps}
      {...others}
    />
  );
};

Tab.defaultProps = {
  "data-id": "tab",
};
Tab.displayName = "Tab";
