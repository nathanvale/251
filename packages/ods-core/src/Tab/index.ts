export { Tab } from "./Tab";
export type TabProps = import("./Tab").TabProps;
