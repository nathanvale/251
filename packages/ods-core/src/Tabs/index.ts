export { Tabs } from "./Tabs";
export type TabsProps = import("./Tabs").TabsProps;
