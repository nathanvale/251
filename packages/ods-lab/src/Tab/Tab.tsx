import React from "react";
import MuiTab, {TabProps as ITab} from "@material-ui/core/Tab";

export interface TabProps extends ITab {}

export const Tab = (props: TabProps) => (
  <MuiTab {...props}>{props.children}</MuiTab>
);
