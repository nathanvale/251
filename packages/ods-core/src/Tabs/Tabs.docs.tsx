import React from "react";
import { ComponentDocs } from "@origin-digital/ods-types";
import { IconGas, IconElectricity } from "@origin-digital/ods-icons";
import { Tab } from "../Tab/Tab";
import { Text } from "../Text/Text";
import { Tabs } from "./Tabs";

export const docs: ComponentDocs = {
  category: "Content",
  componentName: "Tabs",
  description: `A set of components for adding dynamic tab functionality to a page.`,
  migrationGuide: false,
  propDescriptions: {
    defaultValue:
      "If Tabs are uncontrolled, use this prop to set the initial selected Tab",
    value: "Currently selected tab",
    onChange: "Function to be called when a tab is clicked",
    scrollButtons: "",
    children: "Tabs to display - children must be Tab components",
    muiProps: "Allows access to the MUI component API under the hood",
  },
  examples: {
    default: {
      Code: () => (
        <Tabs id="test" defaultValue={0}>
          <Tab data-id="tab1" label="Electricity">
            <Text>Content for electricity</Text>
          </Tab>
          <Tab data-id="tab2" label="Natural gas">
            <Text>Content for natural gas</Text>
          </Tab>
        </Tabs>
      ),
    },
    additional: [
      {
        label: "Controlled tabs",
        description:
          "When you need to reference the tab state from outside the component",
        codeString: `
        const [value, setValue] = React.useState(0);
        const handleChange = (_event, newValue) => setValue(newValue);
  
        return (
          <Tabs id="test" value={value} onChange={handleChange}>
          <Tab data-id="tab1" label="Electricity">
            <Text>Content for electricity</Text>
          </Tab>
          <Tab data-id="tab2" label="Natural gas">
            <Text>Content for natural gas</Text>
          </Tab>
        </Tabs>
          );
        `,
        Code: () => {
          const [value, setValue] = React.useState(0);
          // @ts-ignore
          const handleChange = (_event, newValue) => setValue(newValue);

          return (
            <Tabs id="test" value={value} onChange={handleChange}>
              <Tab data-id="tab1" label="Electricity">
                <Text>Content for electricity</Text>
              </Tab>
              <Tab data-id="tab2" label="Natural gas">
                <Text>Content for natural gas</Text>
              </Tab>
            </Tabs>
          );
        },
        playroom: false,
      },
      {
        label: "Icon tabs",
        description: "Tabs with an icon on them",
        Code: () => (
          <Tabs id="test" aria-label="icon tabs" defaultValue={0}>
            <Tab
              data-id="tab1"
              label="Electricity"
              icon={<IconElectricity tone="neutral" />}
            />
            <Tab
              data-id="tab2"
              label="Natural gas"
              icon={<IconGas tone="neutral" />}
            />
          </Tabs>
        ),
      },
      {
        label: "Scrollable tabs",
        description: "Scroll your tabs",
        Code: () => (
          <Tabs
            id="test"
            aria-label="scrollable tabs"
            defaultValue={0}
            variant="scrollable"
          >
            <Tab
              data-id="tab1"
              label="Electricity"
              icon={<IconElectricity tone="neutral" />}
            />
            <Tab
              data-id="tab3"
              label="Natural gas"
              icon={<IconGas tone="neutral" />}
            />
            <Tab
              data-id="tab4"
              label="Natural gas 2"
              icon={<IconGas tone="neutral" />}
            />
            <Tab
              data-id="tab5"
              label="Natural gas 3"
              icon={<IconGas tone="neutral" />}
            />
            <Tab
              data-id="tab6"
              label="Natural gas 4"
              icon={<IconGas tone="neutral" />}
            />
          </Tabs>
        ),
      },
      {
        label: "Disabled tabs",
        description: "Tabs can be disabled",
        Code: () => (
          <Tabs id="test" defaultValue={0}>
            <Tab
              data-id="tab1"
              label="Electricity"
              icon={<IconElectricity tone="neutral" />}
            />
            <Tab
              data-id="tab2"
              label="Natural gas"
              icon={<IconGas tone="neutral" />}
              disabled
            />
          </Tabs>
        ),
      },
    ],
  },
  snippets: [
    {
      label: "Simple example - uncontrolled",
      Code: () => (
        <Tabs id="test" defaultValue={0}>
          <Tab
            data-id="elec-tab"
            label="Electricity"
            icon={<IconElectricity tone="neutral" />}
          />
          <Tab
            data-id="gas-tab"
            label="Natural gas"
            icon={<IconGas tone="neutral" />}
          />
        </Tabs>
      ),
    },
  ],
};
