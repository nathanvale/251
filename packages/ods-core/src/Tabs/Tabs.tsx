import React from "react";
import {
  Tabs as MuiTabs,
  TabsProps as MuiTabsProps,
  makeStyles,
  Theme,
} from "@material-ui/core";
import { MuiBasedComponentBaseProps, TabSize } from "@origin-digital/ods-types";
import { TabProps } from "../Tab";
import { TabPanel } from "../TabPanel";

export interface TabsProps extends MuiBasedComponentBaseProps {
  "aria-label"?: string;
  value?: string | number;
  defaultValue?: string | number;
  onChange?: (event: React.ChangeEvent<{}>, newValue: string | number) => void;
  variant?: "standard" | "scrollable" | "fullWidth";
  scrollButtons?: "auto" | "on" | "off";
  size?: TabSize;
  hideDivider?: boolean;
  muiProps?: MuiTabsProps;
}

const useTabsStyles = makeStyles(
  (theme: Theme) => ({
    root: {
      borderBottom: ({ hideDivider }: TabsProps) =>
        hideDivider ? "none" : `1px solid ${theme.palette.grey[200]}`,
    },
    indicator: {
      height: "4px",
    },
  }),
  { classNamePrefix: "tabs" }
);

export interface TabsProviderContext {
  size?: TabSize;
}

export const TabsContext = React.createContext<TabsProviderContext>({});

export const Tabs = ({
  defaultValue,
  value,
  onChange,
  hideDivider,
  size,
  children,
  muiProps,
  ...others
}: TabsProps) => {
  const classes = useTabsStyles({ hideDivider });
  const [uncontrolledValue, setUncontrolledValue] = React.useState(
    defaultValue
  );
  const uncontrolledChange = (
    _event: React.ChangeEvent<{}>,
    newValue: string | number
  ) => setUncontrolledValue(newValue);

  const activeTab = value === undefined ? uncontrolledValue : value;

  return (
    <TabsContext.Provider value={{ size }}>
      <MuiTabs
        classes={classes}
        value={activeTab}
        onChange={onChange || uncontrolledChange}
        {...muiProps}
        {...others}
      >
        {children}
      </MuiTabs>
      {React.Children.map(
        children as React.ReactElement<TabProps>[],
        (tab: React.ReactElement<TabProps>, index) => {
          const tabValue = tab.props.value || index;
          return (
            <TabPanel
              hidden={tabValue !== activeTab}
              value={tabValue}
              tabId={tab.props.id}
            >
              {tab.props.children}
            </TabPanel>
          );
        }
      )}
    </TabsContext.Provider>
  );
};

Tabs.defaultProps = {
  scrollButtons: "auto",
  variant: "standard",
  "aria-label": "tabs",
  size: "medium",
};
Tabs.displayName = "Tabs";
