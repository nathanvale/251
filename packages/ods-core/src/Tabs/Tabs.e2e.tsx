import React from "react";
import { E2ETests } from "@origin-digital/ods-types";
import { IconElectricity, IconGas } from "@origin-digital/ods-icons";
import { Tabs } from "../Tabs";
import { Tab } from "../Tab";

export const tests: E2ETests = [
  {
    label: "Default tabs",
    Code: () => (
      <Tabs id="cities" defaultValue="mlb">
        <Tab label="Melbourne" id="mlb" value="mlb">
          Melbourne
        </Tab>
        <Tab label="Sydney" id="syd" value="syd">
          Sydney
        </Tab>
        <Tab label="Perth" id="prt" value="prt">
          Perth
        </Tab>
      </Tabs>
    ),
    responsive: false,
  },
  {
    label: "Icon tabs",
    Code: () => (
      <Tabs id="test" aria-label="icon tabs" defaultValue={0}>
        <Tab
          data-id="tab1"
          label="Electricity"
          icon={<IconElectricity tone="neutral" />}
        />
        <Tab
          data-id="tab2"
          label="Natural gas"
          icon={<IconGas tone="neutral" />}
        />
      </Tabs>
    ),
    responsive: false,
  },
  {
    label: "Disabled tabs",
    Code: () => (
      <Tabs id="test" defaultValue={0}>
        <Tab
          data-id="tab1"
          label="Electricity"
          icon={<IconElectricity tone="neutral" />}
        />
        <Tab
          data-id="tab2"
          label="Natural gas"
          icon={<IconGas tone="neutral" />}
          disabled
        />
      </Tabs>
    ),
    responsive: false,
  },
  {
    label: "Scrollable tabs",
    Code: () => (
      <Tabs
        id="test"
        aria-label="scrollable tabs"
        defaultValue={0}
        variant="scrollable"
      >
        <Tab
          data-id="tab1"
          label="Electricity"
          icon={<IconElectricity tone="neutral" />}
        />
        <Tab
          data-id="tab3"
          label="Natural gas"
          icon={<IconGas tone="neutral" />}
        />
        <Tab
          data-id="tab4"
          label="Natural gas 2"
          icon={<IconGas tone="neutral" />}
        />
        <Tab
          data-id="tab5"
          label="Natural gas 3"
          icon={<IconGas tone="neutral" />}
        />
        <Tab
          data-id="tab6"
          label="Natural gas 4"
          icon={<IconGas tone="neutral" />}
        />
      </Tabs>
    ),
    responsive: false,
  },
];
