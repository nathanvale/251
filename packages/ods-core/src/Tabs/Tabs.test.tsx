import React from "react";
import { render } from "@origin-digital/ods-testing-library";
import { Tab } from "../Tab";
import { Tabs } from "./Tabs";

test("it selects only one Tab in the group when one is clicked", () => {
  const ControlledTabs = () => {
    const [value, setValue] = React.useState("mlb");
    const handleChange = (_e: React.ChangeEvent<{}>, newVal: any) =>
      setValue(newVal);

    return (
      <Tabs id="cities" value={value} onChange={handleChange}>
        <Tab label="Melbourne" id="mlb" value="mlb">
          Melbourne
        </Tab>
        <Tab label="Sydney" id="syd" value="syd">
          Sydney
        </Tab>
        <Tab label="Perth" id="prt" value="prt">
          Perth
        </Tab>
      </Tabs>
    );
  };
  const { container } = render(<ControlledTabs />);

  const [mlbButton, sydButton, prtButton] = Array.from(
    container.getElementsByTagName("button")
  );

  expect(mlbButton.classList).toContain("Mui-selected");
  expect(sydButton.classList).not.toContain("Mui-selected");
  expect(prtButton.classList).not.toContain("Mui-selected");

  sydButton.click();

  expect(mlbButton.classList).not.toContain("Mui-selected");
  expect(sydButton.classList).toContain("Mui-selected");
  expect(prtButton.classList).not.toContain("Mui-selected");
});

test("it renders", () => {
  const { container } = render(
    <Tabs id="cities" value="mlb">
      <Tab label="Melbourne" id="mlb" value="mlb">
        Melbourne
      </Tab>
      <Tab label="Sydney" id="syd" value="syd">
        Sydney
      </Tab>
      <Tab label="Perth" id="prt" value="prt">
        Perth
      </Tab>
    </Tabs>
  );
  expect(container).toMatchSnapshot();
});
