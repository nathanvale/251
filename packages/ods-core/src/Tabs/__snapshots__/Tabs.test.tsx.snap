// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders 1`] = `
.c0 {
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font-size: 100%;
  font: inherit;
  vertical-align: baseline;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c1 {
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font-size: 100%;
  font: inherit;
  vertical-align: baseline;
  margin-top: 32px;
}

<div>
  <div
    class="MuiTabs-root tabs-root-13 tabs-root-15"
    id="cities"
  >
    <div
      class="MuiTabs-scroller MuiTabs-fixed"
      style="overflow: hidden;"
    >
      <div
        aria-label="tabs"
        class="MuiTabs-flexContainer"
        role="tablist"
      >
        <button
          aria-controls="tabpanel-mlb"
          aria-selected="true"
          class="MuiButtonBase-root MuiTab-root tab-root-16 tab-root-18 MuiTab-textColorInherit Mui-selected tab-selected-17"
          data-id="tab"
          id="mlb"
          role="tab"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiTab-wrapper"
          >
            <div
              class="c0"
              display="flex"
            >
              Melbourne
            </div>
          </span>
        </button>
        <button
          aria-controls="tabpanel-syd"
          aria-selected="false"
          class="MuiButtonBase-root MuiTab-root tab-root-16 tab-root-23 MuiTab-textColorInherit"
          data-id="tab"
          id="syd"
          role="tab"
          tabindex="-1"
          type="button"
        >
          <span
            class="MuiTab-wrapper"
          >
            <div
              class="c0"
              display="flex"
            >
              Sydney
            </div>
          </span>
        </button>
        <button
          aria-controls="tabpanel-prt"
          aria-selected="false"
          class="MuiButtonBase-root MuiTab-root tab-root-16 tab-root-24 MuiTab-textColorInherit"
          data-id="tab"
          id="prt"
          role="tab"
          tabindex="-1"
          type="button"
        >
          <span
            class="MuiTab-wrapper"
          >
            <div
              class="c0"
              display="flex"
            >
              Perth
            </div>
          </span>
        </button>
      </div>
      <span
        class="PrivateTabIndicator-root-19 PrivateTabIndicator-colorSecondary-21 MuiTabs-indicator tabs-indicator-14"
        style="left: 0px; width: 0px;"
      />
    </div>
  </div>
  <div
    aria-labelledby="mlb"
    class="c1"
    id="tabpanel-mlb"
    role="tabpanel"
  >
    Melbourne
  </div>
  <div
    aria-labelledby="syd"
    class="c1"
    hidden=""
    id="tabpanel-syd"
    role="tabpanel"
  >
    Sydney
  </div>
  <div
    aria-labelledby="prt"
    class="c1"
    hidden=""
    id="tabpanel-prt"
    role="tabpanel"
  >
    Perth
  </div>
</div>
`;
